File: clients/python/game_objects.py
# -*- python -*-

import operator
import utility
import json
import client_json
import game

class GameObject():
    def __init__(self):
        pass



## @class Player
#  @brief Stores information about a player in the game
class Player(GameObject):

    def __init__(self, connection, parent_game, id, name, byte_dollars, cycles, time):
        self._connection = connection
        self._parent_game = parent_game
        self._id = id
        self._name = name
        self._byte_dollars = byte_dollars
        self._cycles = cycles
        self._time = time

    ## @fn talk
    #  @brief Allows a player to display a message to the screen
    #  @param message The message that the player should say
    def talk(self, message):
        function_call = client_json.function_call.copy()
        function_call.update({"type": 'talk'})
        function_call.get("args").update({"actor": self.id})
        function_call.get("args").update({'message': repr(message)})

        utility.send_string(self.connection, json.dumps(function_call))

        received_status = False
        status = None
        while not received_status:
            message = utility.receive_string(self.connection)
            message = json.loads(message)

            if message.get("type") == "success":
                received_status = True
                status = True
            elif message.get("type") == "failure":
                received_status = True
                status = False
            if message.get("type") == "changes":
                self.parent_game.update_game(message)

        return status

    @property
    def id(self):
        return self._id

    @property
    def name(self):
        return self._name

    @property
    def byte_dollars(self):
        return self._byte_dollars

    @property
    def cycles(self):
        return self._cycles

    @property
    def time(self):
        return self._time




## @class Mappable
#  @brief The base object for all mappable things
class Mappable(GameObject):

    def __init__(self, connection, parent_game, id, x, y):
        self._connection = connection
        self._parent_game = parent_game
        self._id = id
        self._x = x
        self._y = y


    @property
    def id(self):
        return self._id

    @property
    def x(self):
        return self._x

    @property
    def y(self):
        return self._y




## @class Tile
class Tile(Mappable):

    def __init__(self, connection, parent_game, id, x, y, owner):
        self._connection = connection
        self._parent_game = parent_game
        self._id = id
        self._x = x
        self._y = y
        self._owner = owner


    @property
    def id(self):
        return self._id

    @property
    def x(self):
        return self._x

    @property
    def y(self):
        return self._y

    @property
    def owner(self):
        return self._owner




## @class Base
#  @brief The information on the base
class Base(Mappable):

    def __init__(self, connection, parent_game, id, x, y, owner, spawns_left):
        self._connection = connection
        self._parent_game = parent_game
        self._id = id
        self._x = x
        self._y = y
        self._owner = owner
        self._spawns_left = spawns_left

    ## @fn spawn
    #  @brief Creates a Virus on the base with certain level.
    def spawn(self, level):
        function_call = client_json.function_call.copy()
        function_call.update({"type": 'spawn'})
        function_call.get("args").update({"actor": self.id})
        function_call.get("args").update({'level': repr(level)})

        utility.send_string(self.connection, json.dumps(function_call))

        received_status = False
        status = None
        while not received_status:
            message = utility.receive_string(self.connection)
            message = json.loads(message)

            if message.get("type") == "success":
                received_status = True
                status = True
            elif message.get("type") == "failure":
                received_status = True
                status = False
            if message.get("type") == "changes":
                self.parent_game.update_game(message)

        return status

    @property
    def id(self):
        return self._id

    @property
    def x(self):
        return self._x

    @property
    def y(self):
        return self._y

    @property
    def owner(self):
        return self._owner

    @property
    def spawns_left(self):
        return self._spawns_left




## @class Virus
#  @brief Stores the information about a virus
class Virus(Mappable):

    def __init__(self, connection, parent_game, id, x, y, owner, level, moves_left, living):
        self._connection = connection
        self._parent_game = parent_game
        self._id = id
        self._x = x
        self._y = y
        self._owner = owner
        self._level = level
        self._moves_left = moves_left
        self._living = living

    ## @fn move
    #  @param x The x coordinate to move to
    #  @param y The y coordinate to move to
    def move(self, x, y):
        function_call = client_json.function_call.copy()
        function_call.update({"type": 'move'})
        function_call.get("args").update({"actor": self.id})
        function_call.get("args").update({'x': repr(x)})
        function_call.get("args").update({'y': repr(y)})

        utility.send_string(self.connection, json.dumps(function_call))

        received_status = False
        status = None
        while not received_status:
            message = utility.receive_string(self.connection)
            message = json.loads(message)

            if message.get("type") == "success":
                received_status = True
                status = True
            elif message.get("type") == "failure":
                received_status = True
                status = False
            if message.get("type") == "changes":
                self.parent_game.update_game(message)

        return status

    @property
    def id(self):
        return self._id

    @property
    def x(self):
        return self._x

    @property
    def y(self):
        return self._y

    @property
    def owner(self):
        return self._owner

    @property
    def level(self):
        return self._level

    @property
    def moves_left(self):
        return self._moves_left

    @property
    def living(self):
        return self._living

File: clients/cpp/GameObjects.cpp
#include "GameObjects.h"
#include "json/json.h"
#include "Game.h"
#include <sstream>
#include <iostream>
#include <string>

const std::string function_call =
"{\"type\": \"command_name\", \"args\": {\"actor\": 0}}";

Player::Player(GameSocket* connection, Game* parent_game, int id, std::string name, int byte_dollars, int cycles, int time)
{
    this->connection = connection;
    this->parent_game = parent_game;
    this->id = id;
    this->name = name;
    this->byte_dollars = byte_dollars;
    this->cycles = cycles;
    this->time = time;
}

bool Player::talk(std::string message)
{
    std::stringstream convert;
    Json::Value root;
    Json::Reader reader_login;
    reader_login.parse(function_call,root,false);
    root["type"] = "talk";
    root["args"]["actor"] = id;
    convert << message;
    root["args"]["message"] = convert.str();
    convert.str(std::string());

    convert << root;
    //std::cout << root << std::endl;

    if(verbosity == 2)
    {
        std::cout<<"Sent: "<<convert.str()<<'\n';
    }
    connection->send_string(convert.str());
    bool recievedStatus = false;
    bool status;

    while(!recievedStatus)
    {
        std::string message = connection->rec_string();
        Json::Value root2;
        Json::Reader reader_login2;
        reader_login2.parse(function_call,root2,false);
        if(root2["type"] == "success")
        {
            recievedStatus = true;
            status = true;
        }
        else if(root2["type"] == "failure")
        {
            recievedStatus = true;
            status = false;
        }
        else if(root2["type"] == "changes")
        {
            parent_game->update_game(message);
        }
    }
    return status;
}

int Player::get_id()
{
    return id;
}

std::string Player::get_name()
{
    return name;
}

int Player::get_byte_dollars()
{
    return byte_dollars;
}

int Player::get_cycles()
{
    return cycles;
}

int Player::get_time()
{
    return time;
}


Mappable::Mappable(GameSocket* connection, Game* parent_game, int id, int x, int y)
{
    this->connection = connection;
    this->parent_game = parent_game;
    this->id = id;
    this->x = x;
    this->y = y;
}


int Mappable::get_id()
{
    return id;
}

int Mappable::get_x()
{
    return x;
}

int Mappable::get_y()
{
    return y;
}


Tile::Tile(GameSocket* connection, Game* parent_game, int id, int x, int y, int owner)
{
    this->connection = connection;
    this->parent_game = parent_game;
    this->id = id;
    this->x = x;
    this->y = y;
    this->owner = owner;
}


int Tile::get_id()
{
    return id;
}

int Tile::get_x()
{
    return x;
}

int Tile::get_y()
{
    return y;
}

int Tile::get_owner()
{
    return owner;
}


Base::Base(GameSocket* connection, Game* parent_game, int id, int x, int y, int owner, int spawns_left)
{
    this->connection = connection;
    this->parent_game = parent_game;
    this->id = id;
    this->x = x;
    this->y = y;
    this->owner = owner;
    this->spawns_left = spawns_left;
}

bool Base::spawn(int level)
{
    std::stringstream convert;
    Json::Value root;
    Json::Reader reader_login;
    reader_login.parse(function_call,root,false);
    root["type"] = "spawn";
    root["args"]["actor"] = id;
    convert << level;
    root["args"]["level"] = convert.str();
    convert.str(std::string());

    convert << root;
    //std::cout << root << std::endl;

    if(verbosity == 2)
    {
        std::cout<<"Sent: "<<convert.str()<<'\n';
    }
    connection->send_string(convert.str());
    bool recievedStatus = false;
    bool status;

    while(!recievedStatus)
    {
        std::string message = connection->rec_string();
        Json::Value root2;
        Json::Reader reader_login2;
        reader_login2.parse(function_call,root2,false);
        if(root2["type"] == "success")
        {
            recievedStatus = true;
            status = true;
        }
        else if(root2["type"] == "failure")
        {
            recievedStatus = true;
            status = false;
        }
        else if(root2["type"] == "changes")
        {
            parent_game->update_game(message);
        }
    }
    return status;
}

int Base::get_id()
{
    return id;
}

int Base::get_x()
{
    return x;
}

int Base::get_y()
{
    return y;
}

int Base::get_owner()
{
    return owner;
}

int Base::get_spawns_left()
{
    return spawns_left;
}


Virus::Virus(GameSocket* connection, Game* parent_game, int id, int x, int y, int owner, int level, int moves_left, int living)
{
    this->connection = connection;
    this->parent_game = parent_game;
    this->id = id;
    this->x = x;
    this->y = y;
    this->owner = owner;
    this->level = level;
    this->moves_left = moves_left;
    this->living = living;
}

bool Virus::move(int x, int y)
{
    std::stringstream convert;
    Json::Value root;
    Json::Reader reader_login;
    reader_login.parse(function_call,root,false);
    root["type"] = "move";
    root["args"]["actor"] = id;
    convert << x;
    root["args"]["x"] = convert.str();
    convert.str(std::string());
    convert << y;
    root["args"]["y"] = convert.str();
    convert.str(std::string());

    convert << root;
    //std::cout << root << std::endl;

    if(verbosity == 2)
    {
        std::cout<<"Sent: "<<convert.str()<<'\n';
    }
    connection->send_string(convert.str());
    bool recievedStatus = false;
    bool status;

    while(!recievedStatus)
    {
        std::string message = connection->rec_string();
        Json::Value root2;
        Json::Reader reader_login2;
        reader_login2.parse(function_call,root2,false);
        if(root2["type"] == "success")
        {
            recievedStatus = true;
            status = true;
        }
        else if(root2["type"] == "failure")
        {
            recievedStatus = true;
            status = false;
        }
        else if(root2["type"] == "changes")
        {
            parent_game->update_game(message);
        }
    }
    return status;
}

int Virus::get_id()
{
    return id;
}

int Virus::get_x()
{
    return x;
}

int Virus::get_y()
{
    return y;
}

int Virus::get_owner()
{
    return owner;
}

int Virus::get_level()
{
    return level;
}

int Virus::get_moves_left()
{
    return moves_left;
}

int Virus::get_living()
{
    return living;
}



